lib/basedriver/commands/execute.js

```diff
@@ -0,0 +1,102 @@
+import _ from 'lodash';
+import log from '../logger';
+import vm from 'vm';
+import { attach } from 'webdriverio';
+import { MJSONWP_ELEMENT_KEY, W3C_ELEMENT_KEY } from '../../protocol/protocol';
+
+const BASE_PATH = '/wd/hub'; // TODO defining this here is brittle but it is hardcoded in routes, so...
+
+let commands = {};
+
+/*
+ * TODO fill out entire docstring
+ *
+ * This method takes a string which is executed as javascript in the context of
+ * a new nodejs VM, and which has available a webdriverio driver object, having
+ * already been attached to the currently running session.
+ */
+commands.executeDriverScript = async function (script, scriptType) {
+  if (!!scriptType && scriptType !== 'webdriverio') {
+    throw new Error("Only the 'webdriverio' script type is currently supported");
+    // TODO should probably add support for WD script type as well?
+  }
+
+  if (!this.opts.host || !this.opts.port) {
+    throw new Error('Host or port of running server were not defined; this is required');
+  }
+  const driver = attach({
+    sessionId: this.sessionId,
+    protocol: 'http', // Appium won't ever be behind ssl locally
+    hostname: this.opts.host,
+    port: this.opts.port,
+    path: BASE_PATH,
+    isW3C: true, // TODO should probably retrieve client protocol and declare the same one for the script
+    capabilities: this.caps
+  });
+  const fullScript = buildScript(script);
+  const result = await vm.runInNewContext(fullScript)(driver);
+  return coerceScriptResult(result);
+};
+
+function buildScript (script) {
+  return `(async function execute (driver) {
+    ${script}
+  })`;
+}
+
+/*
+ * We can get any manner of crazy thing back from a vm executing untrusted
+ * code. We might also get WebdriverIO objects that aren't suitable for JSON
+ * response. So make sure we convert the things we know about to their
+ * appropriate response format, and squash other weird things.
+ */
+function coerceScriptResult (obj) {
+  // first ensure obj is of a type that can be JSON encoded safely. This will
+  // get rid of custom objects, functions, etc... and turn them into POJOs
+  try {
+    obj = JSON.parse(JSON.stringify(obj));
```

Comment:
this also won't work for bigint number

Reply:
i'm ok documenting that any return value must be consist of simple valid JSON values or elements. if someone is working with bigints they can also do the conversion in the script and return a regular int or some other object representation. i cared about supporting elements as a really common use case.